### ISSUE TITLE
Game crashes after a while with local reference table overflow

### ISSUE THREAD

**Issue by breadbyte** [fexed/Pinball-on-Android#18]
After playing for a minute or two on my phone (Samsung J5 2016), the game crashes.

I am using the [1.2 release.](https://github.com/fexed/Pinball-on-Android/releases/tag/1.2)

Here is a [crash log](https://gist.github.com/breadbyte/a556d2599e80dfd91164b3613f9b0584) spanning multiple games, fetched using `adb logcat --buffer=crash`.

I've noticed that this happens frequently when it's trying to calculate the score after it hits an object.

**Comment by fexed** [comment]
Thanks, will look into this asap!

**Comment by breadbyte** [comment]
I decided to look into the error a bit further.
I got two more logs for two separate games (fetched using `adb logcat --pid=$(adb shell pidof -s com.fexed.spacecadetpinball)`) [[1](https://gist.github.com/breadbyte/e9163dc9bb4ee57ac2fd8be99077d702)] [[2](https://gist.github.com/breadbyte/4d5a6e2d839be10cc7cf2c07f29c2cbc)].

The contents of the reference table for game 2 are as follows:
```
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145] JNI ERROR (app bug): local reference table overflow (max=512)
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145] local reference table dump:
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]   Last 10 entries (of 512):
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       511: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       510: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       509: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       508: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       507: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       506: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       505: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       504: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       503: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       502: 0x12f084c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]   Summary:
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]         1 of java.lang.Thread
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]        34 of java.lang.String (34 unique instances)
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]       477 of java.lang.Class (1 unique instances)
07-02 14:26:03.066 25946 25968 F art     : art/runtime/indirect_reference_table.cc:145]
```

Looking at `SpaceCadetPinballJNI.cpp`, we see the following lines common to all functions in the cpp file:
https://github.com/fexed/Pinball-on-Android/blob/70cc4514906ab515b31847b40fd1c1eaf975f7a6/app/src/main/cpp/SpaceCadetPinballJNI.cpp#L79-L84

Given that there are 477 references to 1 unique class which is `com.fexed.spacecadetpinball.JNIEntryPoint`, this makes me think that
```cpp
jclass clazz = env->FindClass("com/fexed/spacecadetpinball/JNIEntryPoint");
```
keeps the reference and does not drop it after the end of the function, which is confusing because the Android NDK Documentation [says](https://developer.android.com/training/articles/perf-jni#local-and-global-references) that you do not need to clear the reference as it is a local reference and is dropped at the end of the native function, but I digress.

The reason the crash seems to happen often when calculating the score is because, well, you would need to calculate the score after you hit an object, so this function is called often.

This theory is further supported by my [initial crash log](https://gist.github.com/breadbyte/a556d2599e80dfd91164b3613f9b0584), with references to other functions within `SpaceCadetPinballJNI.cpp` that also call `FindClass`.
```
_ZN20SpaceCadetPinballJNI11displayTextEPKci+90 -> SpaceCadetPinballJNI::displayText
_ZN20SpaceCadetPinballJNI9clearTextEi+52 -> SpaceCadetPinballJNI::clearText
_ZN20SpaceCadetPinballJNI9postScoreEi+52 -> SpaceCadetPinballJNI::postScore
_ZN20SpaceCadetPinballJNI16setBallInPlungerEb+52 -> SpaceCadetPinballJNI::setBallInPlunger
```

Dropping this reference after the end of the method invocation or keeping a single global class reference might be the solution for this one.

The Android NDK Documentation [suggests](https://developer.android.com/training/articles/perf-jni#jclass,-jmethodid,-and-jfieldid) keeping a global class reference.

Game 2 also has a similar reference table:
```
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145] JNI ERROR (app bug): local reference table overflow (max=512)
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145] local reference table dump:
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]   Last 10 entries (of 512):
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       511: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       510: 0x12dc4a00 java.lang.String "Reflex Shot Awar... (23 chars)
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       509: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       508: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       507: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       506: 0x12f77310 java.lang.String "Ship Re-Fueled"
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       505: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       504: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.761 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       503: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       502: 0x12efb4c0 java.lang.Class<com.fexed.spacecadetpinball.JNIEntryPoint>
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]   Summary:
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]         1 of java.lang.Thread
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]        42 of java.lang.String (42 unique instances)
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]       469 of java.lang.Class (1 unique instances)
07-02 14:11:29.762 23881 23901 F art     : art/runtime/indirect_reference_table.cc:145]
```

but this time we see `String` references, all of them unique.
Given the contents of the string, `Ship Re-Fueled` and `Reflex Shot Awarded`, this line in `SpaceCadetPinballJNI::displayText`
https://github.com/fexed/Pinball-on-Android/blob/70cc4514906ab515b31847b40fd1c1eaf975f7a6/app/src/main/cpp/SpaceCadetPinballJNI.cpp#L61
may be contributing those string references.

`Reflex Shot Awarded` should be overriding `Ship Re-Fueled`, so it may be necessary to keep tabs on the string references and clear the reference whenever a new string is presented on the screen.

I may be completely wrong on these assumptions, but this is just what I concluded over an hour and a half of googling and looking at code, logs, documentation, and stackoverflow.

I am also unable to test my theories as I do not have the NDK installed currently.

**Comment by fexed** [comment]
This is invaluable insight, thank you. I am quite new to JNI and native code on Android, so this will really help me in figuring this problem out. I'll look into this in the upcoming days and will update here when I have some news!

**Comment by fexed** [comment]
Please let me know if https://github.com/fexed/Pinball-on-Android/releases/tag/1.4 fixes the problem

**Comment by breadbyte** [comment]
Late reply, but it has indeed been fixed by release 1.4 and I can now play for extended periods of time. Thank you! 🎉

**Comment by fexed** [comment]
Glad to hear that and thanks for the help fixing this!
